olp-cpp-sdk-core/src/network2/curl/NetworkProtocolCurl.h

```diff
@@ -0,0 +1,201 @@
+                             std::shared_ptr<std::ostream> payload,
+                             Callback callback,
+                             HeaderCallback header_callback = nullptr,
+                             DataCallback data_callback = nullptr) override;
+
+  void Cancel(RequestId id) override;
+
+ private:
+
+  ErrorCode Send2(const NetworkRequest& request, int id,
+                 const std::shared_ptr<std::ostream>& payload,
+                 Network::HeaderCallback headerCallback,
+                 Network::DataCallback dataCallback,
+                 Network::Callback callback) ;
+
+  bool Initialize();
+
+  void Deinitialize();
+
+  bool Initialized() const;
+
+  bool Ready();
+
+
+  size_t AmountPending();
+
+  void protocolReady();
+
+  static constexpr int kStaticHandleCount = 8;
+  static constexpr int kTotalHandleCount = 32;
+
+  struct RequestHandle {
+    std::string etag{};
+    std::string content_type{};
+    std::string date{};
+    std::chrono::steady_clock::time_point send_time{};
+    std::shared_ptr<std::ostream> payload{};
+    std::weak_ptr<NetworkProtocolCurl> self{};
+    Callback callback{};
+    HeaderCallback header_callback{};
+    DataCallback data_callback{};
+    std::uint64_t count{};
+    std::uint64_t offset{};
+    CURL* handle{nullptr};
+    struct curl_slist* chunk{nullptr};
+    std::uint32_t transfer_timeout{};
+    size_t retry_count{};
+    size_t max_retries{};
+    int index{};
+    int max_Age{};
+    time_t expires{};
+    int id{};
+    bool ignore_offset{};
+    bool in_use{};
+    bool range_out{};
+    bool cancelled{};
+    bool get_statistics{};
+    bool skip_content{};
+    char error_text[CURL_ERROR_SIZE]{};
+  };
+
+  struct EventInfo {
+    enum class Type : char { SendEvent, CancelEvent };
+
+    Type type;
+    RequestHandle* handle;
+
+    EventInfo(Type type, RequestHandle* handle) : type(type), handle(handle) {}
+  };
+
+  static size_t RxFunction(void* ptr, size_t size, size_t nmemb,
+                           RequestHandle* handle);
+  static size_t HeaderFunction(char* ptr, size_t size, size_t nmemb,
+                               RequestHandle* handle);
+
+  int GetHandleIndex(CURL* handle);
+
+  RequestHandle* GetHandle(int id, Network::Callback callback,
+                           Network::HeaderCallback headerCallback,
+                           Network::DataCallback dataCallback,
+                           const std::shared_ptr<std::ostream>& payload);
+  void ReleaseHandle(RequestHandle* handle);
+
+  void ReleaseHandleUnlocked(RequestHandle* handle);
+
+  void CompleteMessage(CURL* handle, CURLcode result);
+
+  /**
+   * @brief The worker thread's main method.
+   */
+  void Run();
+
+  /**
+   * @brief Free resources after the thread terminates.
+   */
+  void Teardown();
+
+  void AddEvent(EventInfo::Type type, RequestHandle* handle);
+
+  RequestHandle handles_[kTotalHandleCount] = { };
+  std::condition_variable event_condition_;
+  std::mutex mutex_;
```

Comment:
A more descriptive name like `event_mutex_` would be nice, currently I have to assume what it's used for.